OrbitCore/SamplingProfiler.cpp

```diff
@@ -287,25 +293,35 @@ void SamplingProfiler::ProcessSamples()
 
     m_State = Processing;
 
+    // collect the callstacks to be processed before hand,
+    //  in order to avoid a dead-lock (HasCallstack also locks).
```

Comment:
Dead-lock with a single mutex?  I think what you are trying to avoid is locking the same mutex twice, however, the Mutex type in Orbit maps to std::recursive_mutex and is reentrant.  There are philosophical debates as to weather that's good or bad, but I find it allows for simpler code.  I'd say acquire the lock once at the top of the function.  ProcessSamples is called once anyways once we've stopped the sampling session, so there's basically no contention.